### PR TITLE
Remove pytrec_eval and rewrite retrieval metrics

### PR DESCRIPTION
**Pull request by cheungdaven**
*Description of changes:*

We removed the dependency on pytrec_eval and rewrite the information retrieval metrics including: precision, recall, ndcg, map, and mrr, with given cutoff values.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: multimodal/src/autogluon/multimodal/utils/metric.py [hunk 1]
```diff
@@ -178,6 +180,171 @@ def compute_score(
...
+            # compute the discounted cumulative gain
+            dcg = np.sum([1 / math.log(rank + 2, 2) for rank in hit_rank])
+
+        # compute the ideal discounted cumulative gain
+        idcg = np.sum([1 / math.log(i + 2, 2) for i in range(at_k)])
+        # compute the normalized discounted cumulative gain
+        ndcg = dcg / idcg
+        mAP /= at_k
+
+        return precision, recall, mrr, mAP, ndcg
```
Comment (sxjscience) [inline multimodal/src/autogluon/multimodal/utils/metric.py:345]: Shall we use dcg_score in sklearn? https://scikit-learn.org/stable/modules/generated/sklearn.metrics.dcg_score.html

Comment (cheungdaven) [reply multimodal/src/autogluon/multimodal/utils/metric.py:345]: dcg_score in sklearn is for classification problems. it requires the input to be "(n_samples, n_labels)" where n_labels is the number of labels, while in ranking tasks, there are no labels and the relevant score are continuous values like cosine similarity.  Also, sklearn requires y_true and y_pred to have the exact same shape, while in our task, we support the same input type as pytrec_eval to ensure minimum code refactor effect in other parts.

File: multimodal/src/autogluon/multimodal/utils/metric.py [hunk 2]
```diff
@@ -178,6 +180,171 @@ def compute_score(
         return metric._sign * metric(metric_data[Y_TRUE], metric_data[Y_PRED])
 
 
+class RankingMetrics:
+    def __init__(
+        self,
+        pred: Dict[str, Dict],
+        target: Dict[str, Dict],
+        is_higher_better=True,
```
Comment (FANGAreNotGnu) [inline multimodal/src/autogluon/multimodal/utils/metric.py:188]: To be consistent with our previous code, here we can call it "min_max_mode". I think "is_higher_better" is easier to understand though.

**Review by FANGAreNotGnu — APPROVED** [review]
LGTM with a minor comment, thanks!

File: multimodal/src/autogluon/multimodal/utils/metric.py (hunk 2 above)
Comment (cheungdaven) [reply multimodal/src/autogluon/multimodal/utils/metric.py:188]: min_max_mode is used for metrics, here is_higher_better is used to indicate how we should rank the list. They are essentially different

**Review by sxjscience — APPROVED** [review]
LGTM

**Review by zhiqiangdon — COMMENTED** [review]
How could we test the implementations?

**Comment by cheungdaven** [comment]
> How could we test the implementations?

I compared the results with pytrec_eval locally. 

For online test, it will be tested in the test cases where you use the "compute_ranking_score" function you added.

**Review by zhiqiangdon — APPROVED** [review]
LGTM! Thanks for the implementations!
